### PR TITLE
update default expressions for C# 7.1

### LINKED ISSUE #2796: Update description of `default` for generics
C# 7.1 adds new syntax for `default` expressions where the type of the expression is inferred.

[This page](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/generics/default-keyword-in-generic-code) needs to be updated to reflect [this feature](https://github.com/dotnet/csharplang/blob/master/proposals/target-typed-default.md)

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #2796

While working on this, I am also addressing some of the text where one could be misled to believe that default() expressions were only valid for Type parameters. The rework should make it clear that default value expressions can be used with any managed type.

### CODE REVIEW

**Comment by BillWagner** [comment]
Opened the PR early because of moving files and checking redirects.

**Comment by BillWagner** [comment]
Proofread and removed the WIP label.  This is ready for review.

@jcouv wrote the language feature specs for this feature. Do you want to review this?

**Review by jcouv — COMMENTED** [review]
Looks good to me. Thanks!

File: docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  - "CSharp"
+helpviewer_keywords: 
+  - "generics [C#], default keyword"
+  - "default keyword [C#], generic programming"
+ms.assetid: b9daf449-4e64-496e-8592-6ed2c8875a98
+caps.latest.revision: 22
+author: "BillWagner"
+ms.author: "wiwagn"
+
+translation.priority.ht:
```
Comment (rpetrusha) [inline docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md]: I'm not sure that these are needed -- at least, when I've included them in new topics, Maira told me to remove them.

File: docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  - "ru-ru"
+  - "tr-tr"
+  - "zh-cn"
+  - "zh-tw"
+---
+# default value expressions (C# programming guide)
+
+A default value expression produces the default value for a type. Default value expressions are particularly useful in generic classes and methods. One issue that arises using generics is how to assign a default value to a parameterized type `T` when you do not know the following in advance:
+
+- Whether `T` will be a reference type or a value type.
```
Comment (rpetrusha) [inline docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md]: nit: the future tense in these two sentences can be replaced with present tense.

File: docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+- If `T` is a value type, whether it will be a numeric value or a user defined struct.
+
+ Given a variable `t` of a parameterized type `T`, the statement `t = null` is only valid if `T` is a reference type and `t = 0` will only work for numeric value types but not for structs. The solution is to use a default value expression, which will return `null` for reference types (class types and interface types) and zero for numeric value types. For user defined structs, it will return the struct initialized to the zero bit pattern, which produces 0 or `null` for each member depending on whether that member is a value or reference types. For nullable value types, default returns a <xref:System.Nullable%601?displayProperty=fullName>, which is initialized like any struct.
```
Comment (rpetrusha) [inline docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md]: type and --> type, and
for `t - 0`, perhaps add "the assignment `t = 0`? (I was going to suggest `=` become `==` until I realized you'd really meant `=`)
a value or reference types --> a value or reference type.
nullable value types, default --> nullable value types, `default`

**Review by rpetrusha — CHANGES_REQUESTED** [review]
This looks good, @BillWagner. Some minor corrections.

File: docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  - "tr-tr"
+  - "zh-cn"
+  - "zh-tw"
+---
+# default value expressions (C# programming guide)
+
+A default value expression produces the default value for a type. Default value expressions are particularly useful in generic classes and methods. One issue that arises using generics is how to assign a default value to a parameterized type `T` when you do not know the following in advance:
+
+- Whether `T` will be a reference type or a value type.
+- If `T` is a value type, whether it will be a numeric value or a user defined struct.
```
Comment (mairaw) [inline docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md]: user defined -> user-defined (Acrolinx catches those. are you running it?)

File: docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md (hunk 3 above)
Comment (mairaw) [inline docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md]: maybe split the sentence at `t = 0` and remove future tense from this paragraph if possible

user defined -> user-defined

File: docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+ [!code-cs[csProgGuideGenerics#1](codesnippet/CSharp/default-value-expressions.cs)]
+
+ The following example from the `GenericList<T>` class shows how to use the `default(T)` operator in a generic class. For more information, see [Generics Overview](../generics/introduction-to-generics.md).
+
+ [!code-cs[csProgGuideGenerics#2](codesnippet/CSharp/default-keyword-in-generic-code_1.cs)]
+
+## default literal and type inference
+
+Beginning with C# 7.1, the `default` literal can be used for default value expressions when the compiler can infer the type of the expression. The `default` literal produces the same value as the equivalent `default(T)` where `T` is the inferred type. This can make code more concise by reducing the redundancy of declaring a type more than once. The `default` literal can be used in any of the following locations:
```
Comment (mairaw) [inline docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md]: the type of the expression -> the expression type

File: docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+- providing the value for a method call argument
+- return statement (or expression in an expression bodied member)
+
+The following example shows many usages of the `default` literal in a default value expression:
+
+[!code-cs[csProgGuideGenerics#3](codesnippet/CSharp/default-literal.cs)]
+
+
+
+## See Also
```
Comment (mairaw) [inline docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md]: lower caps also

File: docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
+---
+title: "default value expressions (C# Programming Guide)"
+
```
Comment (mairaw) [inline docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md]: add description and remove blank lines?

**Review by mairaw — CHANGES_REQUESTED** [review]
LGTM. A few things to address.

File: docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md (hunk 4 above)
Comment (BillWagner) [reply docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md]: I do on new content. I'll start running it on all edits.

File: docs/csharp/programming-guide/statements-expressions-operators/codesnippet/CSharp/default-literal.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
+public class Point
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/programming-guide/statements-expressions-operators/codesnippet/CSharp/default-literal.cs]: I moved the other referenced snippets as well.

**Comment by BillWagner** [comment]
Thanks for the reviews. I updated based on all feedback.

File: docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md [hunk 9]
```diff
@@ -0,0 +1,54 @@
+---
+title: "default value expressions (C# Programming Guide)"
+description: "default value expressions produce the default value for any reference type or value type"
```
Comment (mairaw) [inline docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md]: nit: should this default be capitalized?

File: docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+- return statement (or expression in an expression bodied member)
+
+The following example shows many usages of the `default` literal in a default value expression:
+
+[!code-cs[csProgGuideGenerics#3](../../../../samples/snippets/csharp/programming-guide/statements-expressions-operators/default-literal.cs)]
+
+## See also
+
+ <xref:System.Collections.Generic>
+ [C# Programming Guide](../index.md)
```
Comment (mairaw) [inline docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md]: add the 3 spaces in the end to make the links render on new lines

image

File: docs/csharp/language-reference/keywords/default.md [hunk 11]
```diff
@@ -31,16 +31,19 @@ translation.priority.ht:
...
+
+The `default` keyword can be used in the `switch` statement or in a default value expression:
+
+- [The switch statement](switch.md): Specifies the default label.
+
+- [Default value expressions](../../programming-guide/statements-expressions-operators/default-value-expressions.md): Produces the default value of the type. This will be null for reference types and zero for value types, and the 0 bit pattern for structs.
+
+## See Also
+
+ [C# Reference](../index.md)
```
Comment (mairaw) [inline docs/csharp/language-reference/keywords/default.md]: same here: missing spaces in the end is rendering all links in one line

**Review by mairaw — APPROVED** [review]
LGTM except for a few remaining things I commented on. After fixing those, it's good to merge IMO.

File: docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+caps.latest.revision: 22
+author: "BillWagner"
+ms.author: "wiwagn"
+---
+# default value expressions (C# programming guide)
+
+A default value expression produces the default value for a type. Default value expressions are particularly useful in generic classes and methods. One issue that arises using generics is how to assign a default value to a parameterized type `T` when you do not know the following in advance:
+
+- Whether `T` is a reference type or a value type.
+- If `T` is a value type, whether is a numeric value or a user defined struct.
```
Comment (rpetrusha) [inline docs/csharp/programming-guide/statements-expressions-operators/default-value-expressions.md]: nit: user defined --> user-defined

**Review by rpetrusha — APPROVED** [review]
LGTM, with one nit.

**Pull request merged**